event-statistics/src/main/java/io/quarkus/sample/superheroes/statistics/listener/SuperStats.java

```diff
@@ -1,65 +1,195 @@
-	 */
-	@Incoming("fights")
-	@Outgoing("winner-stats")
-	public Multi<Iterable<Score>> computeTopWinners(Multi<Fight> results) {
-		return results
-			.group().by(Fight::getWinnerName)
-			.flatMap(group ->
-				group.onItem().scan(Score::new, this::incrementScore)
-					.filter(score -> score.getName() != null)
-			)
-			.map(this.topWinners::onNewScore)
-			.invoke(topScores -> LOGGER.debugf("Fight received. Computed the top winners: %s", topScores));
-	}
-
-	private Score incrementScore(Score score, Fight fight) {
-		return new Score(fight.getWinnerName(), score.getScore() + 1);
-	}
-
-	TeamStats getTeamStats() {
-		return this.stats;
-	}
+  private static final Logger LOGGER = Logger.getLogger(SuperStats.class);
+  private static final String SPAN_KEY = "CURRENT_SPAN";
+  private static final String MESSAGE_KEY = "CURRENT_MESSAGE";
+
+  static final String FIGHTS_CHANNEL_NAME = "fights";
+  static final String TEAM_STATS_CHANNEL_NAME = "team-stats";
+  static final String TOP_WINNERS_CHANNEL_NAME = "winner-stats";
+
+  private final Ranking topWinners = new Ranking(10);
+  private final TeamStats stats = new TeamStats();
+
+  private final Tracer tracer;
+  private final MutinyEmitter<TeamScore> teamStatsEmitter;
+
+  public SuperStats(Tracer tracer, @Broadcast @Channel(TEAM_STATS_CHANNEL_NAME) MutinyEmitter<TeamScore> teamStatsEmitter) {
+    this.tracer = tracer;
+    this.teamStatsEmitter = teamStatsEmitter;
+  }
+
+  /**
+   * Processes a stream of {@link Fight}s. Computes {@link #computeTeamStats(io.quarkus.sample.superheroes.fight.schema.Fight, java.util.Optional) team stats}
+   * and {@link #computeTopWinners(io.smallrye.mutiny.Multi) top winners}
+   * @param fights
+   * @return
+   */
+  @Incoming(FIGHTS_CHANNEL_NAME)
+  @Outgoing(TOP_WINNERS_CHANNEL_NAME)
+  @Broadcast
+  public Multi<Message<Iterable<Score>>> processFight(Multi<Message<Fight>> fights) {
+    return fights.withContext((multi, ctx) -> {
+      var teamStatsMulti = createTracedMulti(multi, ctx)
+        .call(fight -> computeTeamStats(fight, getParentSpan(ctx)));
+
+      return computeTopWinners(teamStatsMulti)
+        .onItem().invoke(() -> closeSpanFromContext(ctx))
+        .onFailure().invoke(() -> closeSpanFromContext(ctx))
+        .call(() -> ackMessage(ctx))
+        .map(Message::of);
+    });
+  }
+
+  /**
+   * Transforms the {@link Fight} stream into a stream of {@link io.quarkus.sample.superheroes.statistics.domain.TeamScore scores}.
+   * Each score indicates the running percentage of battles won by heroes.
+   * @param fight The {@link Fight} to compute
+   * @param parentSpan The parent {@link io.opentelemetry.api.trace.Span span}. Used for tracing purposes.
+   * @return A continuous stream of percentages of battles won by heroes sent to the {@code team-stats} in-memory channel.
+   */
+  private Uni<Void> computeTeamStats(Fight fight, Optional<Span> parentSpan) {
```

Comment:
Ahh i think I see what you're saying.

Rather than in a `try`/`catch` block like it is, constuct a `Uni` pipeline that ends the span on both `onItem` or `onFailure` (like I'm doing in `computeTopWinners`)?

Reply:
This seems correct. 

For using `@WithSpan` I tried the following
```java
  @WithSpan("SuperStats.computeTeamStats")
  public Uni<Void> computeTeamStats(@SpanAttribute("arg.fight") Fight fight) {
    LOGGER.debugf("[computeTeamStats] - Got message: %s", fight);
    return Uni.createFrom().item(() -> this.stats.add(fight))
      .onItem().invoke(score -> LOGGER.debugf("[computeTeamStats] - Computed the team statistics: %s", score))
      .onItem().transformToUni(this.teamStatsEmitter::send);
  }
```
and
```java
  private Multi<Fight> createTracedMulti(Multi<Message<Fight>> fights, Context ctx) {
    return fights
      .map(m -> {
        ctx.put(MESSAGE_KEY, m);

        handleTrace(m, "SuperStats.processFight")
          .ifPresent(s -> {
            s.makeCurrent();
            ctx.put(SPAN_KEY, s);
          });

        return m.getPayload();
      });
  }
```

I see that `computeTeamStats` picks the right parent of `processFight` but I can't be sure if `@WithSpan` correctly starts and ends the span for methods returning`Uni`. It is supposed to work with promise-like types according to here https://opentelemetry.io/docs/instrumentation/java/automatic/annotations/#creating-spans-around-methods-with-withspan, but I don't see Uni in the list.

Reply:
I played around with it. It does seem to pick up the `@WithSpan` on the `computeTeamStats` if I make the span current.

The downside is that it forces me to use a non-private method. Following principles of encapsulation, the `computeTeamStats` method should be `private`. But I think making it less complex and leaving it default scope (no access modifier) is an acceptable tradeoff to remove some complexity from the implementation.

Reply:
I just pushed a new commit with this implementation.